aiidalab_widgets_base/computational_resources.py

```diff
@@ -0,0 +1,1033 @@
+    )
+    computers = traitlets.Dict(allow_none=True)
+    allow_select_disabled = traitlets.Bool(False)
+
+    def __init__(self, description="Select computer:", path_to_root="../", **kwargs):
+        """Dropdown for configured AiiDA Computers.
+
+        description (str): Text to display before dropdown.
+
+        path_to_root (str): Path to the app's root folder.
+        """
+
+        self.output = ipw.HTML()
+        self._dropdown = ipw.Dropdown(
+            options={},
+            value=None,
+            description=description,
+            style=STYLE,
+            layout=LAYOUT,
+            disabled=True,
+        )
+        traitlets.link((self, "computers"), (self._dropdown, "options"))
+        traitlets.link((self._dropdown, "value"), (self, "value"))
+
+        self.observe(self.refresh, names="allow_select_disabled")
+
+        children = [
+            ipw.HBox(
+                [
+                    self._dropdown,
+                ]
+            ),
+            self.output,
+        ]
+        self.refresh()
+        super().__init__(children=children, **kwargs)
+
+    def _get_computers(self):
+        """Get the list of available computers."""
+
+        # Getting the current user.
+        user = orm.User.objects.get_default()
+
+        return {
+            c[0].label: c[0]
+            for c in orm.QueryBuilder().append(orm.Computer).all()
+            if c[0].is_user_configured(user)
+            and (self.allow_select_disabled or c[0].is_user_enabled(user))
+        }
+
+    def refresh(self, _=None):
+        """Refresh the list of configured computers."""
+        self.output.value = ""
+        with self.hold_trait_notifications():  # pylint: disable=not-context-manager
+            self._dropdown.options = self._get_computers()
+        if not self.computers:
+            self.output.value = "No computers found."
+            self._dropdown.disabled = True
+        else:
+            self._dropdown.disabled = False
+
+        self._dropdown.value = None
+
+    @traitlets.validate("value")
+    def _validate_value(self, change):
+        """Select computer either by label or by class instance."""
+        computer = change["value"]
+        self.output.value = ""
+        if not computer:
+            return None
+        if isinstance(computer, str):
+            if computer in self.computers:
+                return self.computers[computer]
+            self.output.value = f"""Computer instance '<span style="color:red">{computer}</span>'
+            is not configured in your AiiDA profile."""
+            return None
+
+        if isinstance(computer, orm.Computer):
+            if computer.label in self.computers:
+                return computer
+            self.output.value = f"""Computer instance '<span style="color:red">{computer.label}</span>'
+            is not configured in your AiiDA profile."""
+        return None
+
+
+def CodeDropdown(*args, **kwargs):
+    from warnings import warn
```

Comment:
Any reason to not make this a module-wide import?

Reply:
this has been removed